examples/ddp2pipe/ddp2pipe.py

```diff
@@ -0,0 +1,299 @@
+from pippy.microbatch import TensorChunkSpec, CustomReducer
+from pippy.utils import tp_transports
+
+USE_TQDM = True
+
+DIMS = [28 * 28, 500, 250, 100, 50, 25, 10]
+
+
+class DDPModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mod = nn.Sequential(
+            *[layer for i in range(2) for layer in [nn.Linear(DIMS[i], DIMS[i + 1], bias=True), nn.ReLU()]]
+        )
+
+    def forward(self, x):
+        return self.mod(x)
+
+
+class PipeModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.mod = nn.Sequential(
+            *[layer for i in range(2, len(DIMS) - 1) for layer in
+              [nn.Sequential(nn.Linear(DIMS[i], DIMS[i + 1], bias=True), nn.ReLU())]]
+        )
+
+    def forward(self, x, labels):
+        logits = self.mod(x)
+        loss = F.cross_entropy(logits, labels)
+        return logits, loss
+
+
+class DDP2PipeConnector(nn.Module):
+    def __init__(self, pp_ranks, pp_group, dp_group_size):
+        super().__init__()
+        self.pp_ranks = pp_ranks
+        self.pp_group = pp_group
+
+        rank = torch.distributed.get_rank()
+        driver_rank = self.pp_ranks[0]
+
+        if rank == driver_rank:
+            self.pipe_model = PipeModel()
+
+            for i in range(1, 4):
+                annotate_split_points(self.pipe_model, {f'mod.{i}': PipeSplitWrapper.SplitPoint.BEGINNING})
+
+            # print(self.pipe_model)
+
+            self.pipe = Pipe.from_tracing(self.pipe_model, output_loss_value_spec=(False, True))
+            args_chunk_spec = (TensorChunkSpec(0), TensorChunkSpec(0))
+            kwargs_chunk_spec = {}
+            output_chunk_spec = (TensorChunkSpec(0), CustomReducer(torch.tensor(0.0), lambda a, b: a + b))
+            all_worker_ranks = self.pp_ranks
+            chunks = len(all_worker_ranks)
+
+            # print(self.pipe)
+
+            self.pipeline = PipelineDriverFillDrain(self.pipe, chunks,
+                                                    args_chunk_spec, kwargs_chunk_spec, output_chunk_spec,
+                                                    world_size=len(all_worker_ranks),
+                                                    all_ranks=all_worker_ranks,
+                                                    _debug_mask_minibatches=False)
+
+            self.pipeline.init_data_parallel(dp_group_size)
+
+            self.optimizer = self.pipeline.instantiate_optimizer(optim.Adam)
+
+    def forward(self, x, labels):
+        rank = torch.distributed.get_rank()
+        driver_rank = self.pp_ranks[0]
+
+        if rank == driver_rank:
+            x_list = [torch.empty_like(x) for _ in range(len(self.pp_ranks))]
+            labels_list = [torch.empty_like(labels) for _ in range(len(self.pp_ranks))]
+        else:
+            x_list = None
+            labels_list = None
+
+        torch.distributed.gather(x, gather_list=x_list, dst=driver_rank, group=self.pp_group)
```

Comment:
nice! question: does this support un-even splits?
In general it would be nice to see if we could adopt something like DTensor for this use case later on . c.c. @wanchaol